File: 953-verifying-an-alien-dictionary/953-verifying-an-alien-dictionary.py
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        char_map = {c: i for i, c in enumerate(order)}
        for i in range(1, len(words)):
            for j in range(min(len(words[i]), len(words[i - 1]))):
                if words[i][j] == words[i - 1][j]:
                    continue
                
                if char_map[words[i - 1][j]] > char_map[words[i][j]]:
                    return False
                else:
                    break
            else:
                if len(words[i]) < len(words[i - 1]):
                    return False
        return True
                    

File: 953-verifying-an-alien-dictionary/953-verifying-an-alien-dictionary.java
class Solution {
    public boolean isAlienSorted(String[] words, String order) {
        Map<Character, Integer> charMap = new HashMap<>();
        for(int i = 0; i < order.length(); i++)
            charMap.put(order.charAt(i), i);
        
        for(int i = 1; i < words.length; i++){
            boolean exitEarly = false;
            for(int j = 0; j < Math.min(words[i].length(), words[i - 1].length()); j++){
                if(words[i].charAt(j) == words[i - 1].charAt(j))
                    continue;
                
                if(charMap.get(words[i].charAt(j)) < charMap.get(words[i - 1].charAt(j)))
                    return false;
                else
                    exitEarly = true;
                    break;
            }
            if(!exitEarly && words[i].length() < words[i - 1].length())
                return false;
        }
        
        return true;
    }
}